File: qc-application/src/main/java/com/eager/questioncloud/application/api/post/PostPermissionChecker.java
package com.eager.questioncloud.application.api.post;

import com.eager.questioncloud.core.domain.post.infrastructure.PostRepository;
import com.eager.questioncloud.core.domain.post.model.Post;
import com.eager.questioncloud.core.domain.question.infrastructure.QuestionRepository;
import com.eager.questioncloud.core.domain.userquestion.UserQuestionRepository;
import com.eager.questioncloud.exception.CustomException;
import com.eager.questioncloud.exception.Error;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class PostPermissionChecker {
    private final QuestionRepository questionRepository;
    private final UserQuestionRepository userQuestionRepository;
    private final PostRepository postRepository;

    public Boolean hasPermission(Long userId, Long questionId) {
        if (!questionRepository.isAvailable(questionId)) {
            throw new CustomException(Error.UNAVAILABLE_QUESTION);
        }

        if (!userQuestionRepository.isOwned(userId, questionId)) {
            throw new CustomException(Error.NOT_OWNED_QUESTION);
        }

        return true;
    }

    public Boolean hasCommentPermission(Long userId, Long postId) {
        Post post = postRepository.findById(postId);
        return hasPermission(userId, post.getQuestionId());
    }
}

File: qc-application/src/main/java/com/eager/questioncloud/application/business/post/implement/PostPermissionChecker.kt
package com.eager.questioncloud.application.business.post.implement

import com.eager.questioncloud.core.domain.post.infrastructure.repository.PostRepository
import com.eager.questioncloud.core.domain.question.infrastructure.repository.QuestionRepository
import com.eager.questioncloud.core.domain.userquestion.infrastructure.repository.UserQuestionRepository
import com.eager.questioncloud.core.exception.CoreException
import com.eager.questioncloud.core.exception.Error
import org.springframework.stereotype.Component

@Component
class PostPermissionChecker(
    private val questionRepository: QuestionRepository,
    private val userQuestionRepository: UserQuestionRepository,
    private val postRepository: PostRepository,
) {
    fun hasPermission(userId: Long, questionId: Long): Boolean {
        if (!questionRepository.isAvailable(questionId)) {
            throw CoreException(Error.UNAVAILABLE_QUESTION)
        }

        if (!userQuestionRepository.isOwned(userId, questionId)) {
            throw CoreException(Error.NOT_OWNED_QUESTION)
        }

        return true
    }

    fun hasCommentPermission(userId: Long, postId: Long): Boolean {
        val post = postRepository.findById(postId)
        return hasPermission(userId, post.questionId)
    }
}
